finality-aleph/src/data_io/data_interpreter.rs

```diff
@@ -0,0 +1,132 @@
+use crate::{
+    data_io::{status_provider::ProposalStatusProvider, AlephData},
+    BlockHashNum, SessionBoundaries,
+};
+use async_trait::async_trait;
+use futures::channel::mpsc;
+use log::{debug, error, warn};
+use parking_lot::Mutex;
+use sp_runtime::traits::{Block as BlockT, NumberFor, One, Zero};
+use std::sync::Arc;
+
+/// Takes as input ordered `AlephData` from `AlephBFT` and pushes blocks that should be finalized
+/// to an output channel. The other end of the channel is held by the aggregator whose goal is to
+/// create multisignatures under the finalized blocks.
+pub struct OrderedDataInterpreter<B: BlockT, PSP: ProposalStatusProvider<B>> {
+    blocks_to_finalize_tx: mpsc::UnboundedSender<BlockHashNum<B>>,
+    proposal_status_provider: Arc<Mutex<PSP>>,
+    last_finalized_by_aleph: BlockHashNum<B>,
+    session_boundaries: SessionBoundaries<B>,
+}
+
+impl<B: BlockT, PSP: ProposalStatusProvider<B>> OrderedDataInterpreter<B, PSP> {
+    pub fn new(
+        blocks_to_finalize_tx: mpsc::UnboundedSender<BlockHashNum<B>>,
+        proposal_status_provider: Arc<Mutex<PSP>>,
+        session_boundaries: SessionBoundaries<B>,
+    ) -> Self {
+        let last_finalized_by_aleph = if session_boundaries.first_block() > NumberFor::<B>::zero() {
+            // We are in session > 0, we take the last block of previous session.
+            let last_prev_session_num = session_boundaries.first_block() - NumberFor::<B>::one();
+            proposal_status_provider
+                .lock()
+                .get_finalized_at(last_prev_session_num)
+                .expect("Previous session must have concluded")
```

Comment:
for me it would be kinda confusing to get an error from `OrderedDataInterpreter` because previous session is not concluded :| I would take this logic of getting `last_finalized_by_aleph` out of the constructor of `OrderedDataInterpreter` and just expect here the `BlockHashNum` here

Reply:
I changed the text in the expect, but I'm not sure this logic should be in a different place. This is an invariant that the block must be finalized when the component starts its execution, what we do here is we just explicitly extract this block... Whether we do it here or somewhere else I think it should not matter much.